src/lib/operators/join_hash/join_hash_steps.hpp

```diff
@@ -42,15 +47,68 @@ template <typename T>
 using Partition = std::conditional_t<std::is_trivially_destructible_v<T>, uninitialized_vector<PartitionedElement<T>>,
                                      std::vector<PartitionedElement<T>>>;
 
-// The small_vector holds the first n values in local storage and only resorts to heap storage after that. 1 is chosen
-// as n because in many cases, we join on primary key attributes where by definition we have only one match on the
-// smaller side.
-using SmallPosList = boost::container::small_vector<RowID, 1>;
+// Stores the mapping from HashedType to positions. Conceptually, this is similar to an (unordered_)multimap, but it
+// has some optimizations for the performance-critical probe() method. Instead of storing the matches directly in the
+// hashmap (think map<HashedType, PosList>), we store an offset. This keeps the hashmap small and makes it easier to
+// cache.
+template <typename HashedType>
+class PosHashTable {
+  // In case we consider runtime to be more relevant, the flat hash map performs better (measured to be mostly on par
+  // with bytell hash map and in some cases up to 5% faster) but is significantly larger than the bytell hash map.
+  //
+  // The hash table stores the relative offset of a SmallPosList (see below) in the vector of SmallPosLists. The range
+  // of the offset does not limit the number of rows in the partition but the number of distinct values. If we end up
+  // with a partition that has more values, the partitioning algorithm is at fault.
+  using Offset = uint32_t;
+  using HashTable = ska::bytell_hash_map<HashedType, Offset>;
+
+  // The small_vector holds the first n values in local storage and only resorts to heap storage after that. 1 is chosen
+  // as n because in many cases, we join on primary key attributes where by definition we have only one match on the
+  // smaller side.
+  using SmallPosList = boost::container::small_vector<RowID, 1>;
+
+ public:
+  explicit PosHashTable(JoinHashBuildMode mode, size_t max_size)
+      : _hash_table(static_cast<size_t>(max_size * 1.2)), _pos_lists(max_size), _mode(mode) {
+    // slightly oversize (x 1.2) the hash table to avoid unnecessary rebuilds
```

Comment:
Has anybody ever verified that by experiments?

Reply:
@Bouncner ?

Reply:
This is actually similar to the issue above but still somewhat different. Here, we have already the partitioning in place and just want to have a hash map that "sufficiently" large.

In the other issue, it's about radix partitioning and avoiding that we might end up with hash tables that are larger than the L2 cache.

Reply:
And no: we haven't evaluated it. And I don't know a way to do it properly. It works quite well on the TPC-H (everything is uniform) and can arbitrarily degrade in performance when data is skewed.

Reply:
The factor will be removed. Now that we write much shorter values, the added space should no longer be needed.